Common/src/Protocol/SignallingProtocol.ts

```diff
@@ -90,4 +112,84 @@ export class SignallingProtocol extends EventEmitter {
         this.transport.sendMessage(msg);
         this.transport.emit('out', msg); // emit this for listeners listening to outgoing messages
     }
+
+    private keepalive?: ReturnType<typeof setInterval>;
+    private alive: boolean = false;
+    private initiator: boolean = false;
+
+    private initHeartbeat(): void {
+        // if we're already connected we should receive heartbeats from the initiator
+        // otherwise we're the initiator and should start sending heartbeats
+        this.initiator = !this.transport.isConnected();
+        if (this.initiator) {
+            this.transport.on('open', this.startHeartbeat.bind(this));
+        } else {
+            this.heartbeatWait();
+        }
+    }
+
+    private startHeartbeat(): void {
+        // initiator only.
+        // just sets up an interval to send heartbeats
+        this.alive = true;
+        this.keepalive = setInterval(this.heartbeatInitiate.bind(this), SignallingProtocol.PING_TIMEOUT);
+    }
+
+    private stopHeartbeat(): void {
+        // stops either the heartbeat or the waiting for a heartbeat
+        if (this.initiator) {
+            clearInterval(this.keepalive);
+        } else {
+            clearTimeout(this.keepalive);
+        }
+    }
+
+    private heartbeatInitiate(): void {
+        // initiator only.
+        // if we never got a response from the last heartbeat, assume the connection is dead and timeout
+        if (this.alive === false) {
+            this.disconnect();
+            this.transport.emit('timeout');
+            return;
+        }
+
+        // mark the connection as temporarily dead until we get a response from the ping
+        this.alive = false;
+        this.sendMessage(MessageHelpers.createMessage(Messages.ping, { time: new Date().getTime() }));
+    }
+
+    private heartbeatWait(): void {
+        // non-initiator only.
+        // starts a timer to wait for heartbeat pings. we give it double the ping time to allow some lag
+        this.keepalive = setTimeout(this.heartbeatTimeout.bind(this), SignallingProtocol.PING_TIMEOUT * 2);
```

Comment:
As long as the user can set the value, I feel like it should just timeout to what the users sets, rather than `2*value`.

Additionally, I think if the user sets the timeout to zero or a negative value, then no timeout should be set/fired.